py/frequenz/client/weather/_types.py

```diff
@@ -281,3 +281,85 @@ def to_ndarray_vlf(
             raise RuntimeError("Error processing forecast data") from e
 
         return array
+
+
+@dataclass(frozen=True)
+class HistoricalForecasts:
+    """Historical weather forecast data."""
+
+    _forecasts_pb: weather_pb2.GetHistoricalWeatherForecastResponse
+
+    @classmethod
+    def from_pb(
+        cls, forecasts: weather_pb2.GetHistoricalWeatherForecastResponse
+    ) -> HistoricalForecasts:
+        """Convert a protobuf Forecast message to Forecast object.
+
+        Args:
+            forecasts: protobuf message with historical forecast data.
+
+        Returns:
+            Forecast object corresponding to the protobuf message.
+        """
+        return cls(_forecasts_pb=forecasts)
+
+    # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    def flatten(
+        self,
+    ) -> np.ndarray[
+        tuple[typing.Any, typing.Any, typing.Any, typing.Any, typing.Any, typing.Any],
+        np.dtype[np.float64],
+    ]:
+        """Flatten a Forecast object to a numpy array of tuples of data.
+
+        Returns:
+            Numpy array of tuples with the flattened forecast data.
+
+        Raises:
+            ValueError: If the forecasts data is missing or invalid.
+        """
+        # check for empty forecasts data
+        if not self._forecasts_pb.location_forecasts:
+            raise ValueError("Forecast data is missing or invalid.")
+
+        return flatten(list(self._forecasts_pb.location_forecasts))
+
+
+# pylint: disable=too-many-locals,too-many-branches,too-many-statements
+def flatten(
+    location_forecasts: list[weather_pb2.LocationForecast],
+) -> np.ndarray[
+    tuple[typing.Any, typing.Any, typing.Any, typing.Any, typing.Any, typing.Any],
+    np.dtype[np.float64],
```

Comment:
It seemed a more natural choice to use tuples for me since I did not want to mix objects of different types within the same numpy array. ChatGPT seems to agree as well (https://chat.openai.com/share/4394f8f0-d423-4f92-8504-5a3d14a872b7):

Choosing between using a 2D NumPy array or a NumPy array of tuples depends on your specific use case and the operations you plan to perform. Here are some considerations to help you decide:

**2D NumPy Array**
Use Case: Ideal for mathematical operations, linear algebra, and where you need to perform operations on entire rows or columns simultaneously. If your data represents a matrix or a grid (like in image processing or numerical simulations), a 2D array is usually the way to go.
Performance: Generally offers better performance for vectorized operations, as NumPy is optimized for operations on arrays of uniform data types. This means faster computations and less memory usage compared to a NumPy array of tuples.
Indexing: Allows for more intuitive and flexible indexing and slicing operations, which can be very convenient when working with subarrays or when you need to extract specific rows, columns, or elements.

**NumPy Array of Tuples**
Use Case: Suitable when each element of the array needs to hold a fixed number of elements of potentially different types (e.g., a combination of integers, floats, and strings) and these elements together represent a single entity or record. This structure can be useful for tabular data where each tuple represents a row with different fields.
Flexibility: Provides the flexibility to store elements of different types in each tuple, mimicking structured arrays or records. This can be handy when working with complex data structures that do not fit well into a purely numerical matrix format.
Operations: Operations on arrays of tuples can be less efficient and more cumbersome than on 2D arrays, especially for mathematical computations. You might find yourself iterating over the array more often, which could lead to slower performance compared to vectorized operations on 2D arrays.

**Decision Factors**
Data Homogeneity: If your data is homogeneous (all numerical or all of the same type) and you need to perform mathematical computations, go with a 2D NumPy array.
Data Structure: If your data is structured (like records in a table, each with different types), and you don't need to perform vectorized numerical computations, an array of tuples could be more suitable.
Performance Needs: For high-performance computing needs, especially with large datasets where you'll benefit from vectorization, a 2D array is typically preferred.
In summary, if your primary concern is performance and ease of mathematical computations, a 2D NumPy array is likely the better choice. If you're dealing with complex data structures that require elements of different types to be grouped together, an array of tuples might be more appropriate, albeit with some performance trade-offs.

Reply:
TODO: consider using a named tuple.

Reply:
@cwasicki Replaced numpy arrays with a list of named tuples. Ready for a review again.